File: src/main/java/com/checkmarx/falsepositive/xss/reflected/ReflectedXSS_FP_SafeEncoding.java
package com.checkmarx.falsepositive.xss.reflected;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * FALSE POSITIVE SCENARIOS: Reflected XSS - Safe Encoding Transformations
 * 
 * Pattern: User input encoded/transformed in ways that prevent XSS
 * 
 * These encoding transformations produce output that cannot execute as XSS:
 * - Base64 encoding produces alphanumeric + /+= only
 * - URL encoding escapes special characters
 * - Hex encoding produces only 0-9, A-F
 * 
 * All scenarios are FALSE POSITIVES - SAFE code that should NOT be flagged.
 */
public class ReflectedXSS_FP_SafeEncoding {
    
    /**
     * FALSE POSITIVE: Base64 encoding
     * Base64 output contains only A-Za-z0-9+/= (safe characters)
     */
    public void showBase64Encoded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("data");
        
        // Base64 encoded output cannot contain XSS characters
        String encoded = Base64.getEncoder().encodeToString(userInput.getBytes(StandardCharsets.UTF_8));
        
        PrintWriter out = response.getWriter();
        out.write("Encoded data: " + encoded); // FALSE POSITIVE - Base64 safe alphabet
    }
    
    /**
     * FALSE POSITIVE: URL encoding
     * URLEncoder escapes < > " ' & and other dangerous characters
     */
    public void showUrlEncoded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("query");
        
        // URL encoding escapes XSS characters
        String encoded = URLEncoder.encode(userInput, StandardCharsets.UTF_8.name());
        
        PrintWriter out = response.getWriter();
        out.write("Query: " + encoded); // FALSE POSITIVE - URL encoded
    }
    
    /**
     * FALSE POSITIVE: Hex encoding
     * Hex output is only 0-9, A-F characters
     */
    public void showHexEncoded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("value");
        
        // Hex encoding produces only safe characters
        StringBuilder hex = new StringBuilder();
        for (byte b : userInput.getBytes(StandardCharsets.UTF_8)) {
            hex.append(String.format("%02X", b));
        }
        
        PrintWriter out = response.getWriter();
        out.write("Hex: " + hex.toString()); // FALSE POSITIVE - Hex safe alphabet
    }
    
    /**
     * FALSE POSITIVE: Base64 URL-safe encoding
     * Base64 URL-safe uses - and _ instead of + and /
     */
    public void showBase64UrlEncoded(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("token");
        
        String encoded = Base64.getUrlEncoder().encodeToString(userInput.getBytes(StandardCharsets.UTF_8));
        
        PrintWriter out = response.getWriter();
        out.write("Token: " + encoded); // FALSE POSITIVE - URL-safe Base64
    }
    
    /**
     * FALSE POSITIVE: Integer parsing - output is numeric
     * If parsing succeeds, the value is a pure integer
     */
    public void showParsedInteger(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("id");
        
        try {
            int id = Integer.parseInt(userInput);
            PrintWriter out = response.getWriter();
            out.write("ID: " + id); // FALSE POSITIVE - Numeric only
        } catch (NumberFormatException e) {
            response.getWriter().write("Invalid ID");
        }
    }
    
    /**
     * FALSE POSITIVE: Long parsing - output is numeric
     */
    public void showParsedLong(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("accountId");
        
        try {
            long accountId = Long.parseLong(userInput);
            PrintWriter out = response.getWriter();
            out.write("Account: " + accountId); // FALSE POSITIVE - Numeric only
        } catch (NumberFormatException e) {
            response.getWriter().write("Invalid account ID");
        }
    }
    
    /**
     * FALSE POSITIVE: Double parsing - output is numeric
     */
    public void showParsedDouble(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("price");
        
        try {
            double price = Double.parseDouble(userInput);
            PrintWriter out = response.getWriter();
            out.write("Price: " + price); // FALSE POSITIVE - Numeric only
        } catch (NumberFormatException e) {
            response.getWriter().write("Invalid price");
        }
    }
    
    /**
     * FALSE POSITIVE: Combined encoding transformation
     * Multiple safe transformations
     */
    public void showCombinedEncoding(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("data");
        
        // First Base64, then URL encode (extra safety)
        String base64 = Base64.getEncoder().encodeToString(userInput.getBytes(StandardCharsets.UTF_8));
        String urlEncoded = URLEncoder.encode(base64, StandardCharsets.UTF_8.name());
        
        PrintWriter out = response.getWriter();
        out.write("Safe data: " + urlEncoded); // FALSE POSITIVE - Double encoded
    }
    
    /**
     * FALSE POSITIVE: Hash digest (MD5/SHA)
     * Hash output is only hexadecimal characters
     */
    public void showHashDigest(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String userInput = request.getParameter("value");
        
        try {
            java.security.MessageDigest md = java.security.MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(userInput.getBytes(StandardCharsets.UTF_8));
            StringBuilder hexHash = new StringBuilder();
            for (byte b : hash) {
                hexHash.append(String.format("%02x", b));
            }
            
            PrintWriter out = response.getWriter();
            out.write("Hash: " + hexHash.toString()); // FALSE POSITIVE - Hex only
        } catch (Exception e) {
            response.getWriter().write("Error");
        }
    }
}

File: FalsePositiveTestProject/src/main/csharp/FalsePositive/XSS/ReflectedXSS_FP_SafeEncoding.cs
using System;
using System.Web;
using System.Web.UI;
using System.Text;
using System.Security.Cryptography;

namespace Checkmarx.FalsePositive.XSS
{
    /// <summary>
    /// FALSE POSITIVE SCENARIOS: Reflected XSS - Safe Encoding Transformations (C#)
    /// 
    /// Pattern: User input encoded/transformed in ways that prevent XSS
    /// 
    /// These encoding transformations produce output that cannot execute as XSS:
    /// - Base64 encoding produces alphanumeric + /+= only
    /// - URL encoding escapes special characters
    /// - Numeric parsing produces only numbers
    /// 
    /// All scenarios are FALSE POSITIVES - SAFE code that should NOT be flagged.
    /// </summary>
    public class ReflectedXSS_FP_SafeEncoding : Page
    {
        /// <summary>
        /// FALSE POSITIVE: Base64 encoding
        /// Base64 output contains only A-Za-z0-9+/= (safe characters)
        /// </summary>
        protected void ShowBase64Encoded()
        {
            string userInput = Request.QueryString["data"];
            
            byte[] bytes = Encoding.UTF8.GetBytes(userInput);
            string encoded = Convert.ToBase64String(bytes);
            
            Response.Write("Encoded: " + encoded); // FALSE POSITIVE - Base64 safe
        }
        
        /// <summary>
        /// FALSE POSITIVE: URL encoding
        /// </summary>
        protected void ShowUrlEncoded()
        {
            string userInput = Request.QueryString["query"];
            
            string encoded = HttpUtility.UrlEncode(userInput);
            
            Response.Write("Query: " + encoded); // FALSE POSITIVE - URL encoded
        }
        
        /// <summary>
        /// FALSE POSITIVE: Hex encoding
        /// Hex output is only 0-9, A-F characters
        /// </summary>
        protected void ShowHexEncoded()
        {
            string userInput = Request.QueryString["value"];
            
            byte[] bytes = Encoding.UTF8.GetBytes(userInput);
            string hex = BitConverter.ToString(bytes).Replace("-", "");
            
            Response.Write("Hex: " + hex); // FALSE POSITIVE - Hex safe
        }
        
        /// <summary>
        /// FALSE POSITIVE: int.Parse() - output is numeric
        /// </summary>
        protected void ShowParsedInt()
        {
            string userInput = Request.QueryString["id"];
            
            if (int.TryParse(userInput, out int id))
            {
                Response.Write("ID: " + id); // FALSE POSITIVE - Numeric only
            }
            else
            {
                Response.Write("Invalid ID");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: long.Parse() - output is numeric
        /// </summary>
        protected void ShowParsedLong()
        {
            string userInput = Request.QueryString["accountId"];
            
            if (long.TryParse(userInput, out long accountId))
            {
                Response.Write("Account: " + accountId); // FALSE POSITIVE - Numeric only
            }
            else
            {
                Response.Write("Invalid account");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: decimal.Parse() - output is numeric
        /// </summary>
        protected void ShowParsedDecimal()
        {
            string userInput = Request.QueryString["price"];
            
            if (decimal.TryParse(userInput, out decimal price))
            {
                Response.Write("Price: $" + price); // FALSE POSITIVE - Numeric only
            }
            else
            {
                Response.Write("Invalid price");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: SHA256 hash - output is hex only
        /// </summary>
        protected void ShowHashDigest()
        {
            string userInput = Request.QueryString["value"];
            
            using (SHA256 sha256 = SHA256.Create())
            {
                byte[] hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(userInput));
                string hexHash = BitConverter.ToString(hash).Replace("-", "").ToLower();
                
                Response.Write("Hash: " + hexHash); // FALSE POSITIVE - Hex only
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: GUID parsing - output is GUID format only
        /// </summary>
        protected void ShowParsedGuid()
        {
            string userInput = Request.QueryString["guid"];
            
            if (Guid.TryParse(userInput, out Guid guid))
            {
                Response.Write("GUID: " + guid); // FALSE POSITIVE - GUID format only
            }
            else
            {
                Response.Write("Invalid GUID");
            }
        }
        
        /// <summary>
        /// FALSE POSITIVE: Combined safe transformations
        /// </summary>
        protected void ShowCombinedSafe()
        {
            string userInput = Request.QueryString["data"];
            
            // Base64 then URL encode
            string base64 = Convert.ToBase64String(Encoding.UTF8.GetBytes(userInput));
            string urlEncoded = HttpUtility.UrlEncode(base64);
            
            Response.Write("Safe: " + urlEncoded); // FALSE POSITIVE - Double encoded
        }
    }
}

